Support to string methods (contains, startswith, endswith, equals, isEmpty) to throw NPE (SymbolicStringHandler) GSoC'25

saifk16 opened the pull request:
Earlier there were two choices and null was not being checked:

```java
 if (arg.contains("spf")) {
       System.out.println("Then Side");
 } else {
       System.out.println("Else Side");
 }
```

Let's say we have this if condition then the ***Else Side*** was printed first then ***Then Side*** in the output. But after adding some fixes and one more choice **(choice 0)**, now we get ***NPE*** (if I ignore and try to print in the choice 0, else it will crash) before these two the rest order remains the same. 

Now: 
***choice 0*** -> for the null check
***choice 1*** -> else case
***choice 2*** -> if case

```java
cg = new PCChoiceGenerator(3); // earlier was 2
```

This addon improves the symbolic execution for string method `contains()`

saifk16 committed:
Add three way exploration of symbolic string

saifk16 commented:
https://github.com/SymbolicPathFinder/jpf-symbc/pull/120/commits/5a0170a0c8f4c62f684434ece7004c2666382efe 

Earlier now path condition was added as such, because of I was using sym_v1 rather than sym_v2 (it's a symbolic string -> "string0"), fixed

saifk16 committed:
Handle edge cases in the SymbolicStringHandler for the null choice: GSoC 2025

saifk16 commented:
# Behavior Analysis - https://github.com/SymbolicPathFinder/jpf-symbc/pull/120/commits/96b0b322daadfbd98230dcf45a29da22d28756ac

### Test 1 – Both Strings Concrete
```java
String arg = "HELLO WORLD";
String str = "NACHOS";
```
***Behavior:*** Evaluates at runtime; no symbolic execution involved.

***Output:*** "Else Side" – as arg does not contains str.

***Bytecode Insight:*** Uses ldc to load constants. No symbolic values; no constraint solving needed.

---

### Test 2 – Both Strings Symbolic
```java
String arg = Verifier.nondetString();
String str = Verifier.nondetString();
```
***Behavior:*** Both strings are symbolic; they can take any value including null as well.

***Output:*** NullPointerException

***Reason:*** Symbolic values can be null, and calling contains on null leads to NPE.

***Observation:*** arg and str become symbolic variables (string0, string1).

---

### Test 3 – arg is null (the keyword null which means the reference points to no object), str is symbolic
```java
String arg = null;
String str = Verifier.nondetString();
```
***Behavior:*** Direct method invocation on a null reference.

***Output:*** NullPointerException

***Explanation:*** We cannot call a method using null => `null,contains()`

---

### Test 4 – arg is concrete, str is symbolic
```java
String arg = "Hello";
String str = Verifier.nondetString();
```
***Behavior:*** arg is concrete ("Hello"), str is symbolic (string0).

***Output:*** NullPointerException

***Reason:*** Although arg is valid, str can still be null due to symbolic nature.

---

### Variations
If str = null and arg is symbolic → NPE due to null parameter, as contains() accepts a `@NotNull`

If both arg and str are null → NPE on method call.

for `String str = null`, If we try to do str.contains(arg) => then NPE because you cannot call method on a null reference, when you do arg.contains(str), this is also NPE because contains() accepts `CharSequence`, but in spf/jpf I we do like this sym_v1 is null and sym_v2 is smyblic string0 and when we pop sf.pop() we get a 0.

@sohah

saifk16 committed:
Add 2 choices(EQUALS, NOTEQUAL) => null, in IFNULL bytecodeIns to contradict and execute then and the else side

saifk16 committed:
feat: Add support to IFNONNULL bytecode

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/IFNONNULL.java:
```diff
@@ -3,16 +3,16 @@
  * All rights reserved.
  *
- * Symbolic Pathfinder (jpf-symbc) is licensed under the Apache License, 
+ * Symbolic Pathfinder (jpf-symbc) is licensed under the Apache License,
```
please do not push changes that has to do with formating. This applies in many places in your commits

saifk16 replied:
removed

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/IFNONNULL.java:
```diff
@@ -3,16 +3,16 @@
  * in compliance with the License. You may obtain a copy of the License at
- * 
- *        http://www.apache.org/licenses/LICENSE-2.0. 
+ *
+ *        http://www.apache.org/licenses/LICENSE-2.0.
```
please do not push changes that has to do with formating. This applies in many places in your commits

saifk16 replied:
removed

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/IFNONNULL.java:
```diff
@@ -3,16 +3,16 @@
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
- * See the License for the specific language governing permissions and 
+ * See the License for the specific language governing permissions and
```
please do not push changes that has to do with formating. This applies in many places in your commits

sohah replied:
We cannot accept a pull request with formating changes.

saifk16 replied:
removed

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/SymbolicStringHandler.java:
```diff
@@ -64,6 +64,7 @@
 import gov.nasa.jpf.vm.Types;
 import gov.nasa.jpf.vm.VM;
 import gov.nasa.jpf.vm.StackFrame;
+import gov.nasa.jpf.vm.MJIEnv;
```
please remove this

saifk16 replied:
removed

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/SymbolicStringHandler.java:
```diff
@@ -835,11 +836,13 @@ private void handleBooleanStringInstructions(JVMInvokeInstruction invInst,  Thre
 			assert (cg instanceof PCChoiceGenerator) : "expected PCChoiceGenerator, got: " + cg;
-			conditionValue = (Integer) cg.getNextChoice() == 0 ? false : true;
+			currentChocie = (Integer) cg.getNextChoice();
+			// conditionValue = (Integer) cg.getNextChoice() == 0 ? false : true;
```
You can remove this right?

saifk16 replied:
Removed

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/SymbolicStringHandler.java:
```diff
@@ -864,7 +867,13 @@ private void handleBooleanStringInstructions(JVMInvokeInstruction invInst,  Thre
 
-			if (conditionValue) {
+			if (currentChocie == 2) {
+				System.out.println();
+				System.out.println("============================");
+				System.out.println("When the choice is 2");
+				System.out.println("From the choice 2");
+				System.out.println("============================");
+				System.out.println();
```
Please remove debugging code.

saifk16 replied:
removed

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/SymbolicStringHandler.java:
```diff
@@ -885,7 +894,13 @@ private void handleBooleanStringInstructions(JVMInvokeInstruction invInst,  Thre
 				}
-			} else {
+			} else if(currentChocie == 1) {
+				System.out.println();
+				System.out.println("============================");
+				System.out.println("When choice is 1");
+				System.out.println("From the choice 1");
+				System.out.println("============================");
+				System.out.println();
```
Please remove debugging code.

saifk16 replied:
Removed

sohah reviewed (changes requested):
This pull request does not even compile. Please check. I'm getting compilation error at line 911 in `SymbolicStringHandler.java`. The `conditionValue` is undefined and used in `			sf.push(conditionValue ? 1 : 0, true);`

sohah commented:
> This pull request does not even compile. Please check. I'm getting compilation error at line 911 in `SymbolicStringHandler.java`. The `conditionValue` is undefined and used in ` sf.push(conditionValue ? 1 : 0, true);`

I think I was wrong about this comment. I'll recheck that PR

saifk16 commented:
> > This pull request does not even compile. Please check. I'm getting compilation error at line 911 in `SymbolicStringHandler.java`. The `conditionValue` is undefined and used in ` sf.push(conditionValue ? 1 : 0, true);`
> 
> I think I was wrong about this comment. I'll recheck that PR

Latest push the commit 71a8d03 compiles on my pc, I check and compile before pushing I am working on the issue. Thanks for the comments

sohah commented:
@saifk16 , can you please write the test cases under the examples folder, along with their .jpf file. Having them as part of the description of the PR is great, but we do want to use these tests to check the changes, and for future regression tests. Thank you.

saifk16 commented:
> @saifk16 , can you please write the test cases under the examples folder, along with their .jpf file. Having them as part of the description of the PR is great, but we do want to use these tests to check the changes, and for future regression tests. Thank you.

Okay ma'am

sohah commented:
> > @saifk16 , can you please write the test cases under the examples folder, along with their .jpf file. Having them as part of the description of the PR is great, but we do want to use these tests to check the changes, and for future regression tests. Thank you.
> 
> Okay ma'am

Thank you, please make make a single file with the different variations, one for testing the IFNULL and another for testing the IFNONNULL. I'll look at this pull request, you the tests are added.

saifk16 commented:
59c4009 till this commit my code complies

saifk16 commented:
[9ab5e89](https://github.com/SymbolicPathFinder/jpf-symbc/pull/120/commits/9ab5e89ba5e9126f9755bae57697be71c71aab4e) Till this commit my code compiles successfully.

saifk16 committed:
Add support isEmpty refactor

saifk16 committed:
Add test cases for testing npe (contains)

saifk16 committed:
Refactor npe.jpf

saifk16 committed:
Add trim support, roll back charAt and length

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/SymbolicStringHandler.java:
```diff
@@ -900,14 +911,38 @@ private void handleBooleanStringInstructions(JVMInvokeInstruction invInst,  Thre
+							pc.spc._addDet(StringComparator.EQUALS, sym_v1, "null");
+						}
+					} else {
+						pc.spc._addDet(StringComparator.EQUALS, sym_v2, "null");
```
You should not make this, because if we do `str.contains(arg)`, where str is concrete while `arg` is symbolic, no null pointer exception should be raised. Here you should just ignore this check and not make any constraint.

saifk16 replied:
Hello ma'am, I totally respect your decision and understand what you are saying, if `arg` is **symbolic** then it can be **null** as well if we call a concrete value of `str` using `contains` on a **null literal** then it will raise a `null pointer exception` because contains has parameters annotated `@NotNull` which means they should not be null as the internal implementation of contains uses tostring() so null.toString() will throw npe as we know that we cannot call a null on a method, I tested it out will it be the correct behavior or I should just ignore ?

Here arg is symbolic so it can be null and str is concrete, hance we call null on toString() method when we pass a symoblic arg (which can be null) to contains method which accepts CharSequence.

saifk16 replied:
Hello @sohah ma'am I was successfully able to check when the method we invoke is equals and a concrete call using equals on the symbolic value is made when it can be null. Currently I am ignoring it, I have added all the test cases to the methods I supported and also ran them on my machine, they work as they should be when the `runtime.exception` is on and off.

sohah commented on jpf-symbc/src/examples/npe/StringContainsTest.java:
```diff
@@ -0,0 +1,276 @@
+
+    // Test 8: if str is not null - branch check (arg != null) => (IFNULL)
+    // Branching occurs because str is symbolic and can be either null or not null
+    public static void testNotNull1() {
```
Please seperate the testing of the IFNULL into a seperate test file. Same for the test for IFNONNULL.

saifk16 replied:
Done

sohah commented on jpf-symbc/src/examples/npe/npe.jpf:
```diff
@@ -0,0 +1,19 @@
+target=npe.StringContainsTest
```
where is the flag for the `npe` here?

saifk16 replied:
I did not added that in this file (runtime.exception=true or runtime.exception=false)

saifk16 replied:
it was added later

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/IFNONNULL.java:
```diff
@@ -3,16 +3,16 @@
  * All rights reserved.
  *
- * Symbolic Pathfinder (jpf-symbc) is licensed under the Apache License, 
+ * Symbolic Pathfinder (jpf-symbc) is licensed under the Apache License,
```
Please do not add formatting changes with your pull request, i.e., changes in spaces, etc should be ignored and not part of the pull request. We should only see changes in the code.

saifk16 replied:
Sorry, I didn't mean to do that, I will check from now onwards carefully.

sohah replied:
Please fix, you cannot push changes like these, i.e., formatting related. Please fix.

saifk16 replied:
this is removed

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/SymbolicStringHandler.java:
```diff
@@ -835,13 +844,15 @@ private void handleBooleanStringInstructions(JVMInvokeInstruction invInst,  Thre
+
+			Config conf = th.getVM().getConfig();
+			String[] re = conf.getStringArray("runtime.exception");
+			boolean re_flag = re != null && re[0].equalsIgnoreCase("true");
```
This should be a single field within the SymbolicStringHandler, not with every method. Here you should just check it.

saifk16 replied:
If I make this a single field then I will have to create a constructor for this, also if the flag is not specified in the .jpf file by the user then either it should give null pointer exception or it should skip the npe check as the flag will be false by default.

saifk16 replied:
This was made in a single field

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/SymbolicStringHandler.java:
```diff
@@ -1185,19 +1220,48 @@ public void handleTrim(JVMInvokeInstruction invInst, ThreadInfo th) {
+			if (!pc.simplify()) {
+				th.getVM().getSystemState().setIgnored(true);
+			} else {
+				th.createAndThrowException("java.lang.NullPointerException");
```
all these check should be checked if your `npe` flag is true, we only add then the null checks to the pc and check them, otherwise we ignore. this should happen to all functions where we raise those expections similar to what we have done for the `contains`. It is not something that is special to contains only, all string functions that could raise a null pointer exception should include this check.

saifk16 replied:
done

sohah reviewed (changes requested):
Please update, and please do not add new features (more support ) to this pull request, until we can get it in. But of course, please fix and address its comments. 
And also, you can make local changes with new support, but keep them local to you, until we get this pull request in. Thank you!

saifk16 committed:
Add final commit(code) with test cases for contains

saifk16 commented:
# 71a9319
Till this commit I have added the support for checking null choice in the SymbolicStringHandler.
Supported methods:

All the below methods are supported under the `handleBooleanStringInstructions` method.
All these methods will have `three` choices one where it will be checked for null and the other two then and the else case, like contains or does not contains.

1. contains()
2. startsWith()
3. endsWith()
4. equals()
---
5. isEmpty()
method responsible `handleIsEmpty`
This will also have `three` choices one for null and the other two to check whether empty or not empty.
---

`IFNULL` and `IFNONNULL` bytecode support was added for when user specifies






```java
arg != null and arg == null
// arg here is symbolic
```
A flag was also introduced `runtime.exception` with `true/false` if this is enabled the null choice is explored if not then it is skipped. Single field was needed in the `SymbolicStringHandler` so needed to create a constructor for it and pass `ThreadInfo th` to it then also passed the `th` in `bytecodeutils`.

For now only test cases for contains() are added in three different files (standard/ifnull/ifnonnull).

saifk16 commented:
# 02af087 
Added test cases to the methods supported in this pr till now (contains, equals, startsWith, endsWith, isEmpty), also handled a specail case for equals.

---
when we call a `str.equals(arg)` where `str` is concrete and `arg` is smybolic and if it then it can be `null` as well, so equals will return false for this case and no NPE will be raised because equals in String class is for content comparison. Rest of the methods like contains, startsWith and endsWith will raise NPE in this case because of their internal workings and semantics.

sohah commented:
HI @saifk16 , what is the status of this pull request, have you included the changes we recently talked about for the "String.equals"?

saifk16 committed:
Refactor SymbolicStringHandler for equals

saifk16 commented:
> HI @saifk16 , what is the status of this pull request, have you included the changes we recently talked about for the "String.equals"?

yes ma'am just now sorry for the inconvenience caused

a9cfdfd

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/IFNONNULL.java:
```diff
@@ -46,20 +51,69 @@
+							((PCChoiceGenerator) cg).setCurrentPC(pc);
+						}
+						return getTarget();
+					}
```
this should only be enable if you have the npe flag turned on.

saifk16 replied:
done

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/IFNONNULL.java:
```diff
@@ -46,20 +51,69 @@
+				}
+			} else {
+				return ti.createAndThrowException("java.lang.UnsupportedOperationException", "IFNONNULL for non-string symbolic expressions is not supported.");
+			}
```
this too needs to be wrapped with the if-condition that checks for the npe flag turned on

saifk16 replied:
done

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/IFNULL.java:
```diff
@@ -3,16 +3,16 @@
  *
  * Unless required by applicable law or agreed to in writing, software
  * distributed under the License is distributed on an "AS IS" BASIS,
  * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
```
please remove formatting changes

saifk16 replied:
removed

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/IFNULL.java:
```diff
@@ -47,20 +52,68 @@
 	public IFNULL (int targetPc) {
-	    super(targetPc);
-	  }
+		super(targetPc);
```
too many format changes. please fix

saifk16 replied:
removed

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/IFNULL.java:
```diff
@@ -47,20 +52,68 @@
+					}
+				}
+			} else {
+				return ti.createAndThrowException("java.lang.UnsupportedOperationException", "IFNULL for non-string expressions is not supported.");
```
please wrap this code with the check for the npe turned on

saifk16 replied:
done

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/SymbolicStringHandler.java:
```diff
@@ -1195,9 +1257,9 @@ public void handleTrim(JVMInvokeInstruction invInst, ThreadInfo th) {
-																																 */
+		 * dummy String
+		 * Object
+		 */
```
format changes.

saifk16 replied:
done, removed.

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/SymbolicStringHandler.java:
```diff
@@ -897,20 +912,67 @@ private void handleBooleanStringInstructions(JVMInvokeInstruction invInst,  Thre
+						String val = e1.asString();
+						pc.spc._addDet(comp.not(), val, sym_v2);
+					} else {
+						pc.spc._addDet(comp.not(), "null", sym_v2);
```
why are you doing this, when the currentChoice==1?

saifk16 replied:
This is done because of these lines 
```java
ElementInfo e1 = th.getElementInfo(s1);
String val = e1.asString();
```
if we use arg -> symbolic and str -> concrete null literal then e1.asString() gives null pointer exception

sohah replied:
You need to leave this code as before. You should care about null values only in the null choice, which is choice number 0, right? You should not be changing in other choices.

saifk16 replied:
This case was when the string that is checked for equals, contains, startswith, endswith like str.contains(string), so the string can be concrete null, at that time e1.asString() gives error in the output, if npe flag will be off so the choice 0 for null pointer check will be skipped, and the two choices will be 1 and 2 that will e1.asString() as soon as we pass a concrete null it will give error.

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/SymbolicStringHandler.java:
```diff
@@ -875,8 +886,12 @@ private void handleBooleanStringInstructions(JVMInvokeInstruction invInst,  Thre
+						String val = e1.asString();
+						pc.spc._addDet(comp, val, sym_v2);
+					} else {
+						pc.spc._addDet(comp, "null", sym_v2);
```
here too, why are you doing this? why are you checking for "null" for choice==2?

saifk16 replied:
This is done because of these lines
```java
ElementInfo e1 = th.getElementInfo(s1);
String val = e1.asString();
```
if we use arg -> symbolic and str -> concrete null literal then e1.asString() gives null pointer exception

saifk16 replied:
This case was when the string that is checked for equals, contains, startswith, endswith like str.contains(string), so the string can be concrete null, at that time e1.asString() gives error in the output, if npe flag will be off so the choice 0 for null pointer check will be skipped, and the two choices will be 1 and 2 that will e1.asString() as soon as we pass a concrete null it will give error.

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/SymbolicStringHandler.java:
```diff
@@ -897,20 +912,67 @@ private void handleBooleanStringInstructions(JVMInvokeInstruction invInst,  Thre
+				if (!re_flag) {
+					th.getVM().getSystemState().setIgnored(true);
+				} else {
+					if(isEqualsMethod) {
```
I think it would be better if you nested this within the particular check, i.e., body of the if-statement rather than as an outer filtering condition. This is because if you make the check on the other level you'll basically have to repeate a big block of the code when you can just filter at the right time, i.e., when you are about to check the "arg"  symbolic value for null

saifk16 replied:
sorry for this change I have made this work correctly in the recent commit, it should be like this
re flag -> if true then ignore else go the null check, then first we check if the method is eqauls or not if yes then we do the equals thing else we go normally as it should be done for the rest of the methods

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/SymbolicStringHandler.java:
```diff
@@ -1329,23 +1391,38 @@ public void handleIsEmpty(JVMInvokeInstruction invInst,  ThreadInfo th) {
+				sf.push(0, true);
+			} else if (conditionValue == 2) {
+				sf.push(1, true);
+			}
```
i do not think this is needed, we can push something on the sf too in the case we want to do setIgnored(true). It doesn't really matter what we put on the sf, as we are going to ignore this path anyway. So, i think we can leave the code as is.

saifk16 replied:
earlier in the code when cloned from the sv-comp branch this code was:
```java
sf.push(conditionValue ? 1 : 0, true);
```
so I thought now when isEmpty has three choices just like the contains, equals, startswith, endswith and isEmpty is also a boolean based method so I should push as I was in the `handleBooleanStringInstructions` method

saifk16 replied:
this was removed, done.

saifk16 replied:
[StackFrame.java](https://github.com/javapathfinder/jpf-core/blob/45a4450cd0bd1193df5419f7c9d9b89807d00db6/src/main/gov/nasa/jpf/vm/StackFrame.java#L576) 

I get a assertion error for this when the nullpointerexception is set to false.

sohah reviewed (changes requested):
Hi @saifk16 , here are some comments, please address them one by one, and please leave a comment on the particular code-snippet where I left you the comments. That way, we can easily track our discussions. Thank you.

saifk16 committed:
Refactor SymbolicStringHandler for the null choice in handleBooleanStringInstructions

saifk16 commented:
d6490be - till this commit most of the reviews have been finished only the sf push and the npe flag wrap around IFNULL and IFNONNULL is left, need to test them more.

saifk16 committed:
Refactor and introduce new flag with the script additions

saifk16 commented:
@sohah ma'am, every comment or review on this pr till now has been finished as I added and refactored and fixed the things asked, I also pushed the updated script (svcomp) as well. 

The pr is again available for the next review, THANKS.

saifk16 committed:
Fix one error in svcomp script

saifk16 committed:
Refactor SymbolicStringHandler

saifk16 committed:
Refactored svcomp script

saifk16 commented:
script changed as asked, and the sf.push as well. 5019af8

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/IFNONNULL.java:
```diff
@@ -46,20 +52,78 @@
+                        } else {
+                            pc.spc._addDet(StringComparator.NOTEQUALS, (StringExpression) sym_v, "null");
+                            if (!pc.simplify()) {
+                                ti.getVM().getSystemState().setIgnored(true);
```
Please flip the choices so that the then-side is executed first, then the else-side after it.

saifk16 replied:
done

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/IFNULL.java:
```diff
@@ -54,13 +59,70 @@ public Instruction execute (ThreadInfo ti) {
+                        boolean currentChoice = (Integer) cg.getNextChoice() == 0;
+
+                        // two choices (EQUALS, "null") | (NOTEQUALS, "null")
+                        if (currentChoice) {
+                            pc.spc._addDet(StringComparator.EQUALS, (StringExpression) sym_v, "null");
+                            if (!pc.simplify()) {
+                                ti.getVM().getSystemState().setIgnored(true);
+                            } else {
+                                ((PCChoiceGenerator) cg).setCurrentPC(pc);
+                            }
+                            return getTarget();
+                        } else {
+                            pc.spc._addDet(StringComparator.NOTEQUALS, (StringExpression) sym_v, "null");
+                            if (!pc.simplify()) {
+                                ti.getVM().getSystemState().setIgnored(true);
+                            } else {
+                                ((PCChoiceGenerator) cg).setCurrentPC(pc);
+                            }
+                            return getNext(ti);
+                        }
+                    }
+                } else {
```
Please flip the choices so that the then-side is executed first, then the else-side after it.

saifk16 replied:
done

sohah reviewed (changes requested):
@saifk16, the PR looks that it is in a more stable state than before, but still needs some refinements. please check my commits and please address them.

Also, can you please remove all "String**IfNonNullTest" and "String**IfNullTest" and only have a SINGLE test file, that have checks for both null and non-null, without the using "contains or equals" as these are different functionality. Your test can use if-null and if-non-null.

saifk16 committed:
Choices flipped for the IFNULL and IFNONNULL and also moved the test in one file

saifk16 commented:
> @saifk16, the PR looks that it is in a more stable state than before, but still needs some refinements. please check my commits and please address them.
> 
> Also, can you please remove all "String**IfNonNullTest" and "String**IfNullTest" and only have a SINGLE test file, that have checks for both null and non-null, without the using "contains or equals" as these are different functionality. Your test can use if-null and if-non-null.

@sohah ma'am the choices are flipped and the ifnull and the ifnonnull test case files are removed and we now have only one file which has the test cases for ifnull and ifnonnull.

saifk16 committed:
listener flag in the script

saifk16 committed:
else case when npe flag is off and a concrete null is passed

saifk16 commented:
@sohah ma'am the else part https://github.com/SymbolicPathFinder/jpf-symbc/pull/120#discussion_r2289981066 for the above review is done.

sohah commented on jpf-symbc/src/main/gov/nasa/jpf/symbc/bytecode/SymbolicStringHandler.java:
```diff
@@ -875,8 +890,21 @@ private void handleBooleanStringInstructions(JVMInvokeInstruction invInst,  Thre
+                        System.err.println("\nWe do not expect a concrete null to be used when the nullPointer.exception flag is set to false.");
+                        System.err.println("At - " + th.getTopFrame().getClassInfo().getName() + ":" + th.getTopFrame().getLine());
+
+                        System.exit(1);
```
Please replace with

`System.out.println("warning: unexpected concrete null arguments are detected")`

no exit, no errors, just a text warning

saifk16 replied:
Done.

sohah reviewed (changes requested):
Can you please address this one comment. Thank you.

saifk16 commented:
@sohah ma'am I did the change that was asked by you. Thank you.

sohah commented:
Great! Thanks @saifk16 .
@yannicnoller , can you please accept this PR. I think it is ready to be accepted. Also, Saif needs to create another PR based on this one, once it is accepted. Thank you!

saifk16 commented:
@sohah ma'am and @yannicnoller sir, thanks for reviewing and merging.
